src/internal/hash.c

```diff
@@ -246,3 +259,149 @@ void hash_skip(struct hash_state_t* p_state, void* p_input, int n) {
 #endif // USE_VERSION_3
 
 #undef USE_VERSION_3
+
+// -----------------------------------------------------------------------------
+
+r_obj* ffi_hash_file(r_obj* path) {
+  XXH3_state_t* p_xx_state = XXH3_createState();
+
+  struct exec_data data = {
+    .x = path,
+    .p_xx_state = p_xx_state
+  };
+
+  return R_ExecWithCleanup(hash_file_impl, &data, hash_cleanup, &data);
+}
+
+#define CHUNK_SIZE 512 * 1024
+
+static
+r_obj* hash_file_impl(void* p_data) {
+  struct exec_data* p_exec_data = (struct exec_data*) p_data;
+  r_obj* path = p_exec_data->x;
+  XXH3_state_t* p_xx_state = p_exec_data->p_xx_state;
+
+  if (!r_is_character(path, -1)) {
```

Comment:
`r_is_character()` is no longer part of the C lib and mainly remains because it is used to implement `purrr::is_character()`. It should probably be renamed with a different prefix.

We normally use `r_typeof(path) != R_TYPE_character` for checking types.